pkg/git/git.go

```diff
@@ -43,42 +43,57 @@ func run(logger *zap.SugaredLogger, dir string, args ...string) (string, error)
 	return output.String(), nil
 }
 
+// FetchSpec describes how to initialize and fetch from a Git repository.
+type FetchSpec struct {
+	URL      string
+	Revision string
+	Path     string
+	Depth    uint
+}
+
 // Fetch fetches the specified git repository at the revision into path.
-func Fetch(logger *zap.SugaredLogger, revision, path, url string) error {
+func Fetch(logger *zap.SugaredLogger, spec FetchSpec) error {
 	if err := ensureHomeEnv(logger); err != nil {
 		return err
 	}
 
-	if revision == "" {
-		revision = "master"
+	if spec.Revision == "" {
+		spec.Revision = "master"
 	}
-	if path != "" {
-		if _, err := run(logger, "", "init", path); err != nil {
+	if spec.Path != "" {
+		if _, err := run(logger, "", "init", spec.Path); err != nil {
 			return err
 		}
-		if err := os.Chdir(path); err != nil {
-			return fmt.Errorf("Failed to change directory with path %s; err: %w", path, err)
+		if err := os.Chdir(spec.Path); err != nil {
+			return fmt.Errorf("Failed to change directory with path %s; err: %w", spec.Path, err)
 		}
 	} else if _, err := run(logger, "", "init"); err != nil {
 		return err
 	}
-	trimmedURL := strings.TrimSpace(url)
+	trimmedURL := strings.TrimSpace(spec.URL)
 	if _, err := run(logger, "", "remote", "add", "origin", trimmedURL); err != nil {
 		return err
 	}
-	if _, err := run(logger, "", "fetch", "--depth=1", "--recurse-submodules=yes", "origin", revision); err != nil {
+
+	fetchArgs := []string{"fetch", "--recurse-submodules=yes"}
+	if spec.Depth > 0 {
```

Comment:
Ah I see, we just ignore any value <= 0. This seems like it could surprise users, I think we should reject it upfront instead. 🤔 

Or at least log that we're ignoring it? Also probably worth logging that we couldn't parse the depth as a uint, and that we're falling back to the default.

Reply:
> I think we should reject [depth of 0]

The original issue asking for depth control (#1450) isn't super clearly written, but my interpretation is that the author was asking for a way to turn off shallow clones entirely:

> actual behavior: Git resources are always shallow cloned

`git fetch --depth 0` is an error, so I’m overloading it to mean “full depth”. But I agree it’s subtle; how about I add a separate param `shallow` (defaulting to `true`) as the mechanism to opt into a full clone?

> Also probably worth logging that we couldn't parse the depth as a uint, and that we're falling back to the default.

Unrecognized values for `submodules` are currently swallowed without logging anything – I kept this consistent with what’s already there, but I agree it’s not user-friendly. What if we have `NewGitResource` return an error in these cases? Would that be triggered from the validating webhook?

Reply:
Yeah that should bubble up to webhook validation.

FWIW I think this change is pretty much just fine without all this extra validation, if you want to get this in and fix the obvious regression first, then add better validation later.

Reply:
that would be fine by me 😎